File: lv1_module2/ros2_ws/src/turtle_cpp/src/distance_subscriber.cpp
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/float32.hpp>

class DistanceSubscriber : public rclcpp::Node
{
public:
    DistanceSubscriber() : Node("distance_subscriber_cpp")
    {
        dist_sub_ = this->create_subscription<std_msgs::msg::Float32>(
            "/turtle_distance", 10,
            std::bind(&DistanceSubscriber::dist_callback, this,
                      std::placeholders::_1));

        RCLCPP_INFO(this->get_logger(), "C++ Distance subscriber started");
    }

private:
    void dist_callback(const std_msgs::msg::Float32::SharedPtr msg)
    {
        RCLCPP_INFO(this->get_logger(), "distance=%.2f", msg->data);
    }

    rclcpp::Subscription<std_msgs::msg::Float32>::SharedPtr dist_sub_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<DistanceSubscriber>());
    rclcpp::shutdown();
    return 0;
}

File: lv1_module2/ros2_ws/src/turtle_py/turtle_py/distance_subscriber.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import Float32


class DistanceSubscriber(Node):
    def __init__(self):
        super().__init__('distance_subscriber')

        self.declare_parameter('warn_distance', 2.5)

        self.dist_sub = self.create_subscription(
            Float32, '/turtle_distance', self.dist_callback, 10)

        self.get_logger().info('Distance subscriber started')

    def dist_callback(self, msg):
        """거리값을 받아 임계값 초과 시 경고 로그"""
        threshold = self.get_parameter('warn_distance').value

        if msg.data > threshold:
            self.get_logger().warn(
                f'Turtle too far from origin! '
                f'distance={msg.data:.2f}, threshold={threshold:.1f}')
        else:
            self.get_logger().info(f'distance={msg.data:.2f}')


def main(args=None):
    rclpy.init(args=args)
    node = DistanceSubscriber()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()
